dir.c

```diff
@@ -1387,25 +1387,37 @@ static enum path_treatment treat_directory(struct dir_struct *dir,
 		return path_none;
 
 	case index_nonexistent:
-		if (dir->flags & DIR_SHOW_OTHER_DIRECTORIES)
-			break;
+		if (dir->flags & DIR_SHOW_OTHER_DIRECTORIES) {
+			if (!(dir->flags & DIR_HIDE_EMPTY_DIRECTORIES))
+				return exclude ? path_excluded : path_untracked;
+
+			untracked = lookup_untracked(dir->untracked, untracked,
+				     dirname + baselen, len - baselen);
+			return read_directory_recursive(dir, istate, dirname, len,
+					untracked, 1, 0, pathspec);
+		}
+		if (exclude &&
+			(dir->flags & DIR_SHOW_IGNORED_TOO) &&
+			(dir->flags & DIR_SHOW_IGNORED_DIRECTORY)) {
```

Comment:
I wonder whether it would make sense to include `DIR_SHOW_IGNORED_TOO` in `DIR_SHOW_IGNORED_DIRECTORY`, i.e. define `DIR_SHOW_IGNORED_DIRECTORY` as `1 << 7 | 1 << 8`... Then it would be obvious that you cannot use the latter without the former... What do you think?